docs/architecture/adr-042-state-sync.md

```diff
@@ -0,0 +1,225 @@
+* A Set of MultiStore APIs for exposing snapshot data to the ABCI - ABCI application
+* A Storage format with validation and performance considerations - ABCI application
+
+### Implementation Properties
+Beyond the approach, any implementation of StateSync can be evaluated
+across different criteria:
+
+* Speed: Expected throughput of producing and consuming snapshots
+* Safety: Cost of pushing invalid snapshots to a node
+* Liveness: Cost of preventing a node from receiving/constructing a snapshot
+* Effort: How much effort does an implementation require
+
+### Implementation Question
+
+* What is the format of a snapshot
+    * Complete snapshot 
+    * Ordered IAVL key ranges
+    * Compressed individually chunks which can be validated
+* How is data validated
+    * Trust a peer with it's data blindly
+    * Trust a majority of peers
+    * Use light client validation to validate each chunk against consensus
+      produced merkle tree root
+* What are the performance characteristics
+    * Random vs sequential reads
+    * How parallelizeable is the scheduling algorithm
+
+### Proposals
+Broadly speaking there are two approaches to this problem which have had
+varying degrees of discussion and progress. These approach can be
+summarized as:
+
+**Lazy:** Where snapshots are produced dynamically at request time. This
+solution would use the existing data structure.
+**Eager:** Where snapshots are produced periodically and served from disk at
+request time. This solution would create an auxiliary data structure
+optimized for batch read/writes.
+
+#### Lazy StateSync
+An [initial specification](https://docs.google.com/document/d/15MFsQtNA0MGBv7F096FFWRDzQ1vR6_dics5Y49vF8JU/edit?ts=5a0f3629) was published by Alexis Sellier.
+In this design, the state has a given `size` of primitive elements (like
+keys or nodes), each element is assigned a number from 0 to `size-1`,
+and chunks consists of a range of such elements.  Ackratos raised
+[some concerns](https://docs.google.com/document/d/1npGTAa1qxe8EQZ1wG0a0Sip9t5oX2vYZNUDwr_LVRR4/edit)
+about this design, somewhat specific to the IAVL tree, and mainly concerning
+performance of random reads and of iterating through the tree to determine element numbers
+(ie. elements aren't indexed by the element number).
+
+This design could be generalized to have numbered chunks instead of
+elements, where chunks are indexed by their chunk number and determined
+by some algorithm for grouping the elements in the tree. This would
+allow pre-computation and persistence of the chunks that could alleviate
+the concerns about real-time iteration and random reads.  However, there
+seems to be a larger issue with the "numbered chunks" approach: it does
+not seem possible, in general, to verify that the chunk corresponds to
+the chunk number, even if it can be verified that the chunk is a valid
+set of nodes in the tree. A malicious peer could provide a chunk that
```

Comment:
What if there is a deterministic function from subsets of key spaces to the natural numbers (chunk numbers)?

Then the downloading node can check the keys they've received in the chunk match the number.

It requires a bit of prior agreement on the function but I don't see why that wouldn't work?

It could even vary; e.g. the client could pick a modulus `n` and the chunk could be the hash of the key mod `n`.

Reply:
I suspect this conversation continued in #3799